CEP XXXX: Customizable system DLL linkage checks for windows (#128)
conda/ceps: cep-XXXX.md

   where `<rpath>` entries are listed in `build/rpath`), the DSO has to be installed as
   well; either as part of the same package or one of its dependencies.

2. If the DSO was in a sysroot (`$BUILD_PREFIX/*/sysroot`), then these are system
   libraries and are allowed. These DSOs usually come from `sysroot_<subdir>`
   conda packages.

For macOS,

1. If the DSO was in the specified host environment locations (`$PREFIX/<rpath>`,
   where `<rpath>` entries are listed in `build/rpath`), the DSO has to be installed as
   well; either as part of the same package or one of its dependencies.

2. If the DSO was in `CONDA_BUILD_SYSROOT` (the macOS SDK location), then these are system
   libraries and are allowed.

For Windows,

1. If the DSO was in the specified host environment locations (`$PREFIX/Scripts`,
   `$PREFIX/Library/bin`, etc), the DSO has to be installed as
   well; either as part of the same package or one of its dependencies.

2. If the DSO was in `CONDA_BUILD_SYSROOT` (this is unset in most scenarios), then   these are system libraries and are allowed.

3. If the DSO was in `C:\Windows\System32`, then these are system libraries and are
   allowed.

> 4. If the DSO was in the `DEFAULT_WIN_WHITELIST` location defined by the build tool,

Reviewer: `DEFAULT_WIN_ALLOWLIST`?

Reviewer: Sorry, missed that this was describing existing behavior in conda-build.

are not always found in user systems. Also, linking to these debug DLLs
are not desirable and therefore should not be shipped it the corresponding
conda package (e.g. `vc14_runtime`). Therefore we need a way to tell the build tool
which DLLs are allowed and which are not.

## Specification

This CEP proposes a mechanism to control the allowed system DLLs on Windows.
When a file provided by a conda package in `host` or `build` exists in any of
the following locations we use those list of files:

- `<BUILD_PREFIX>/etc/conda-build/(allow|deny)list.d/<subdir>/<conda-package-name>.txt`

- `<PREFIX>/etc/conda-build/(allow|deny)list.d/<subdir>/<conda-package-name>.txt`

When no allow list is available, but only a deny list is available, a default `C:\Windows\System32\*.dll` glob is assumed.
When no deny list is available and an allow list is available,
an empty deny list is assumed. When neither are available,
conda build tools should fall back to the current method (as of August 2025).

Each list supports standard POSIX globbing syntax; i.e., the following is allowed:

- `C:\Windows\System32\*.dll`
- `C:\foo.dll`
- `**\R.dll`

> Note that backslashes are allowed and only absolute paths are allowed.

Reviewer 2: Do we expect any files in the $PREFIX to be whitelisted? One could allow relative paths, or $PREFIX substitution.

Reviewer 2: Backslashes + globs might be a dangerous combination because we cannot escape characters then I suppose?

However, the user has no way of controlling which system DSO are allowed
on windows.

### Need for customizability of system DLLs

A user of conda-build/rattler-build might need to customize the set of system DLLs.
An example from Linux is

  [Removal of libnsl from glibc](https://github.com/conda-forge/linux-sysroot-feedstock/pull/40)

In this case newer versions of glibc dropped libnsl and made it a separate package,
therefore conda-forge dropped it from `sysroot_linux-*` packages to make sure
when new conda packages links to libnsl system DLL, conda-build/rattler-build will
warn/error out.

An example for Windows is,

  [Warn when linking to debug VC runtimes](https://github.com/conda/conda-build/issues/5732)

In this case, on the CI machines we have the debug VC runtime DLLs and are
considered system DLLs by conda-build/rattler-build, but these debug DLLs
are not always found in user systems. Also, linking to these debug DLLs
are not desirable and therefore we do not ship it in our `vc14_runtime`
conda package. Therefore we need a way to tell conda-build/rattler-build
on which DLLs are desirable and which are not.

> ## Specification

Reviewer 3: This section should probably mention that these files are expected to be provided by conda packages depended on in `build` or `host`.

In this CEP, it is proposed that we allow a user to control the allowed
system DLLs on windows. When a file exists in the following locations
we use those list of files:

   <BUILD_PREFIX>/etc/conda-build/(allow|deny)list.d/<subdir>/<conda-package-name>.txt

   <PREFIX>/etc/conda-build/(allow|deny)list.d/<subdir>/<conda-package-name>.txt

When no allow list is available, but only a deny list is available, a default

   C:\Windows\System32\*.dll

is assumed. When no deny list is available and an allow list is available,
an empty deny list is assumed. When neither are available,
conda-build/rattler-build should fall back to the current method.

Each list supports globbing syntax, i.e., following is allowed.

   C:\Windows\System32\*.dll
   C:\foo.dll
   **\R.dll

## Conclusion

> In this CEP, we propose a method to allow users of conda-build/rattler-build

Reviewer 3: ```suggestion
In this CEP, we propose a method to allow users of conda build tools
```

> to control the allowed system DLLs on windows, making windows have feature

Reviewer 3: ```suggestion
to control the allowed system DLLs on Windows, bringing it to feature
```

which DLLs are allowed and which are not.

## Specification

This CEP proposes a mechanism to control the allowed system DLLs on Windows.
When a JSON file provided by a conda package in `host` or `build` exists in
any of the following locations we use those list of files:

- `<BUILD_PREFIX>/etc/conda-build/dsolists.d/<conda-package-name>.json`

- `<PREFIX>/etc/conda-build/dsolists.d/<conda-package-name>.json`

The JSON file format is:

```json
{
  "version": 1,
  "allow": [str],
  "deny": [str],
  "subdir": str
}
```

Here `version` is for the version of the DSO list JSON schema and will always
be present for these files. The build tools should check the version and
validate that the `version` is a known and supported version and error out
if not.

The elements of the two lists `allow` and `deny` supports standard POSIX
globbing syntax; i.e., the following is allowed:

```json
{
  "version": 1,
  "allow": [
    "C:/Windows/System32/*.dll",
    "**/R.dll",
  ],
  "deny": [
    "**/ucrtbased.dll",
  ],
  "subdir": "win-64",
}
```

Note that forward slashes are used for path separation and only absolute paths

> are allowed.

Reviewer 4: Are these rules (only absolute paths, standard POSIX globbing syntax) supposed to be checked by the build tools?

## Specification

This CEP proposes a mechanism to control the allowed system DLLs on Windows.
When a JSON file provided by a conda package in `host` or `build` exists in
any of the following locations we use those list of files:

- `<BUILD_PREFIX>/etc/conda-build/dsolists.d/<conda-package-name>.json`

- `<PREFIX>/etc/conda-build/dsolists.d/<conda-package-name>.json`

The JSON file format is:

```json
{
  "version": 1,
  "allow": [str],
  "deny": [str],
  "subdir": str
}
```

Here `version` is for the version of the DSO list JSON schema and will always
be present for these files. The build tools should check the version and
validate that the `version` is a known and supported version and error out
if not. This CEP specifies `version=1` only and subsequent CEPs can revise
this format by changing the `version` key.

The elements of the two lists `allow` and `deny` supports standard POSIX
globbing syntax; i.e., the following is allowed:

```json
{
  "version": 1,
  "allow": [
    "C:/Windows/System32/*.dll",
    "**/R.dll",
  ],
  "deny": [
    "**/ucrtbased.dll",
  ],
  "subdir": "win-64",
}
```

Note that forward slashes are used for path separation and only absolute paths

> are allowed. This rule should be enforced by the build tool.

Reviewer 3: I guess the `**/something` globs are considered absolute too. Can you confirm and, in that case, clarify the text to cover them too? Thanks!
